### DISCUSSION TITLE
How do I add a shift in the offsets of a las file and output into another las file?

### DISCUSSION THREAD

**Discussion by imaginaryBuddy** [laspy/laspy#320] [Q&A]
Using the following codes, it doesn't seem to be working as my output file only has Z values when I open it up on QGIS. 

Could someone guide me on this? Thank you! 

```
import laspy
# Shift in latitude and longitude
shift_lat = 103.83964917 - 103.83963837
shift_lon = 1.35651542 - 1.35653912

# Paths to input and output LAS files
input_file = <input>
output_file =<output>


input_las = laspy.read(input_file)
print("old offsets: ")
print(input_las.header.offsets)


output_header = input_las.header
output = laspy.LasData(output_header)
output.X = input_las.X
output.Y = input_las.Y 
output.Z = input_las.Z 

output.xyz = input_las.xyz

output.header.offsets = [
  input_las.header.offsets[0] + shift_lat,
  input_las.header.offsets[1] + shift_lon, 
  input_las.header.offsets[2] 
]

output.write(output_file)

print("new offsets: ")
print(output.header.offsets)

```

**Comment by tmontaigu** [comment]
Just changing the header is not good enoug

you have to use https://laspy.readthedocs.io/en/latest/api/laspy.lasdata.html#laspy.lasdata.LasData.change_scaling

**Comment by imaginaryBuddy** [comment]
@tmontaigu, thank you very much! 

I also found out that I have to calculate the offsets first, and it cannot be directly computed and assigned in 1 expression:

```
import laspy 

# Shift in latitude and longitude
shift_lat = 103.83964917 - 103.83963837
shift_lon = 1.35651542 - 1.35653912

# Paths to input and output LAS files
input_file = <input>
output_file = <output>

input_las = laspy.read(input_file)

x_offset = input_las.header.offsets[0] + shift_lat
y_offset = input_las.header.offsets[1] + shift_lon

new_offsets = [input_las.header.offsets[0] + shift_lat,
               input_las.header.offsets[1] + shift_lon,
               input_las.header.offsets[2]]

''' this doesn't work 
new_offsets = [ input_las.header.offsets[0] + shift_lat,
               y_offset,
               input_las.header.offsets[2]]
'''

input_las.change_scaling(offsets=new_offsets)


input_las.write(output_file)
```

**Comment by imaginaryBuddy** [comment]
Actually, I would like to have some clarification on how actually to shift the by a certain lat, lon. 

Because, if I take two points (A and B), where B is the destination point that A is supposed to offset to, I am unable to shift the lidar file (I know this because the points don't shift after running the code on QGIS). However, the offset of the output file changes accordingly, and the offset in the header is changed as well. 

I am unsure why this happens, do you know why?

**Reply by tmontaigu** [reply]
I'm not sure what you mean by shifting the lidar file. Do you mean actually moving the points ?

The offset/scale mechanism of LAS file is intended to preserve accuracy of the data as the file stores points using integers not floating points
